src/lib/communication/DY.Lib.Communication.API.Predicates.fst

```diff
@@ -0,0 +1,138 @@
+open DY.Lib.Event.Typed
+
+open DY.Lib.Communication.API
+
+#set-options "--fuel 0 --ifuel 0 --z3cliopt 'smt.qi.eager_threshold=100'"
+
+(*** PkEnc Predicates ***)
+
+#push-options "--ifuel 3 --fuel 0"
+val pkenc_crypto_predicates_communication_layer: cusages:crypto_usages -> pkenc_crypto_predicate cusages
+let pkenc_crypto_predicates_communication_layer cusages = {
+  pred = (fun tr pk msg ->
+    get_sk_usage pk == PkKey comm_layer_pkenc_tag empty /\
+      (match parse communication_message msg with
+      | Some (CM {sender; receiver; payload}) -> (
+        get_sk_label pk == principal_label receiver /\
+        (get_label payload) `can_flow tr` (join (principal_label sender) (principal_label receiver)) /\
+        event_triggered tr sender (CommConfSendMsg sender receiver payload)
+      )
+      | _ -> False)
+    );
+  pred_later = (fun tr1 tr2 pk msg -> ());
+}
+#pop-options
+
+val pkenc_crypto_predicates_communication_layer_and_tag: 
+  {|cusages:crypto_usages|} ->
+  (string & pkenc_crypto_predicate cusages)
+let pkenc_crypto_predicates_communication_layer_and_tag #cusages = 
+  (comm_layer_pkenc_tag, pkenc_crypto_predicates_communication_layer cusages)
+
+(*** Sign Predicates ***)
+
+#push-options "--ifuel 3 --fuel 0"
+val sign_crypto_predicates_communication_layer: cusages:crypto_usages -> sign_crypto_predicate cusages
+let sign_crypto_predicates_communication_layer cusages = {
+  pred = (fun tr vk sig_msg ->
+    get_signkey_usage vk == SigKey comm_layer_sign_tag empty /\
+    (match parse communication_message sig_msg with
+    | Some (CM {sender; receiver; payload}) -> (
+      get_signkey_label vk == principal_label sender /\
+      get_label payload `can_flow tr` public /\
+      event_triggered tr sender (CommAuthSendMsg sender payload)
+    )
+    | _ -> False)
+  );
+  pred_later = (fun tr1 tr2 vk msg -> ());
+}
+#pop-options
+
+val sign_crypto_predicates_communication_layer_and_tag: 
+  {|cusages:crypto_usages|} ->
+  (string & sign_crypto_predicate cusages)
+let sign_crypto_predicates_communication_layer_and_tag #cusages =
+  (comm_layer_sign_tag, sign_crypto_predicates_communication_layer cusages)
+
+val has_communication_layer_invariants:
+  crypto_invariants ->
+  prop
+let has_communication_layer_invariants cinvs =
+  has_pkenc_predicate cinvs pkenc_crypto_predicates_communication_layer_and_tag /\
+  has_sign_predicate cinvs sign_crypto_predicates_communication_layer_and_tag
+
+(*** Event Predicates ***)
+
+noeq
+type comm_higher_layer_event_preds = {
+  send_conf: sender:principal -> receiver:principal -> payload:bytes -> tr:trace -> prop;
+  // TODO do we need extensibility on the receiver side?
+  receive_conf: sender:principal -> receiver:principal -> payload:bytes -> tr:trace -> prop;
+  send_auth: sender:principal -> payload:bytes -> tr:trace -> prop;
+  // TODO do we need extensibility on the receiver side?
+  receive_auth: sender:principal -> receiver:principal -> payload:bytes -> tr:trace -> prop;
+}
```

Comment:
What are the purpose of `receive_conf` and `receive_auth`? `send_conf` and `send_auth` are useful to transmit properties from the sender to the receiver, but `receive_conf` and `receive_auth` can only transmit properties from the receiver to itself?